### PR TITLE
usm: http2: Support large payloads

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Handle an edge case when a request or a response contains large body.
Before the PR we supported max of 10 frames per packet, while large bodies can have more frames in a single packet.
The new logic is filters up to 500 frames, and looks for at most 10 interesting frames (HEADERS frames, or data frames with end of stream flag).

Then we parse the interesting frames as before.

### Motivation

Supporting requests with large bodies.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 1]
```diff
@@ -305,8 +305,41 @@ static __always_inline void process_headers_frame(struct __sk_buff *skb, http2_s
...
-    // Checking for preface
+static __always_inline void parse_frame(struct __sk_buff *skb, skb_info_t *skb_info, conn_tuple_t *tup, http2_ctx_t *http2_ctx, struct http2_frame *current_frame) {
+    http2_ctx->http2_stream_key.stream_id = current_frame->stream_id;
+    http2_stream_t *current_stream = http2_fetch_stream(&http2_ctx->http2_stream_key);
+    if (current_stream == NULL) {
+        skb_info->data_off += current_frame->length;
+        return;
+    }
+
+    if (current_frame->type == kHeadersFrame) { // This is a header frame.
```
Comment (amitslavin) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:316]: Do you think that the doc is relevant? is not adding any info

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 2]
```diff
@@ -305,8 +305,41 @@ static __always_inline void process_headers_frame(struct __sk_buff *skb, http2_s
...
+        process_headers_frame(skb, current_stream, skb_info, tup, &http2_ctx->dynamic_index, current_frame);
+    } else {
+        skb_info->data_off += current_frame->length;
+    }
+
+    if ((current_frame->flags & HTTP2_END_OF_STREAM) == HTTP2_END_OF_STREAM) { // This is end of stream.
```
Comment (amitslavin) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:322]: Same for this doc, what is the added value of it? even the variable names say the same

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 3]
```diff
@@ -305,8 +305,41 @@ static __always_inline void process_headers_frame(struct __sk_buff *skb, http2_s
...
+        handle_end_of_stream(current_stream, &http2_ctx->http2_stream_key);
+    }
+
+    return;
+}
+
+static __always_inline bool format_http2_frame_header(struct http2_frame *out) {
```
Comment (amitslavin) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:329]: can you please add doc for the function, and in general add doc for other functions the file as well,  I do think we should improve some doc of this file in some places

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 4]
```diff
@@ -315,71 +348,55 @@ static __always_inline bool http2_entrypoint(struct __sk_buff *skb, skb_info_t *
...
+        bpf_skb_load_bytes(skb, skb_info->data_off, (char*)&current_frame, HTTP2_FRAME_HEADER_SIZE);
+        skb_info->data_off += HTTP2_FRAME_HEADER_SIZE;
+        if (!format_http2_frame_header(&current_frame)){
+            break;
+        }
 
-    if (is_end_of_stream) {
-        handle_end_of_stream(current_stream, &http2_ctx->http2_stream_key);
+        is_headers_frame = current_frame.type == kHeadersFrame;
+        is_data_end_of_stream = ((current_frame.flags & HTTP2_END_OF_STREAM) == HTTP2_END_OF_STREAM) && (current_frame.type == kDataFrame);
```
Comment (amitslavin) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:378]: add doc to the RFC to validate that those conditions are correct for the end of stream for data frame

File: pkg/network/protocols/grpc/monitor_test.go [hunk 5]
```diff
@@ -221,14 +221,13 @@ func testGRPCScenarios(t *testing.T) {
 					Method: http.MethodPost,
 				}: 5,
 			},
-			expectedError: true,
 		},
 		{
-			name: "request with large body (1MB) -> b -> request with large body (1MB) -> b",
+			name: "request with large body (5MB) -> b -> request with large body (5MB) -> b",
```
Comment (amitslavin) [inline pkg/network/protocols/grpc/monitor_test.go:226]: do we know what is the current max body size that we can support? I can understand that it is probably related to the gRPC max size, but maybe it is worth to use some UTs to understand our max size

**Review by amitslavin — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/grpc/monitor_test.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/protocols/grpc/monitor_test.go:226]: unlimited, that depends on the how the http2 is split into packets
I'll run a UT with 1G

File: pkg/network/ebpf/c/protocols/http2/decoding.h (hunk 1 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:316]: removed

File: pkg/network/ebpf/c/protocols/http2/decoding.h (hunk 2 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:322]: removed

File: pkg/network/ebpf/c/protocols/http2/decoding.h (hunk 3 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:329]: done

File: pkg/network/ebpf/c/protocols/http2/decoding.h (hunk 4 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:378]: done

**Review by amitslavin — APPROVED** [review]

**Pull request merged**